### DISCUSSION TITLE
`Ref` of wrong type in generated class

### DISCUSSION THREAD

**Discussion by 2x2xplz** [rrousselGit/riverpod#3225] [Q&A] [answered]
Attempting to migrate non-codegen providers to codegen. Some providers rely on `ref.watch()`ing other providers. Following #1697 I replaced `Ref` with `AutoDisposeRef` as a `build()` parameter in my annotated classes. This eliminated `The argument type '...' can't be assigned to the parameter type 'AlwaysAliveProviderListenable<dynamic>'.` errors inside the build() function.

However the generated abstract classes have an issue with the `ref` parameter:

`_$ActiveFoos.ref' ('AutoDisposeRef<dynamic> Function()') isn't a valid override of 'BuildlessAutoDisposeStreamNotifier.ref' ('AutoDisposeStreamNotifierProviderRef<Set<Foo>> Function()')`

A very simple version of my class is:
```
@riverpod
class ActiveFoos extends _$ActiveFoos {
  @override Stream<Set<Foo>> build(AutoDisposeRef ref) async* {
    yield <Foo>[].toSet();
    // ... real class listens to another StreamProvider via ref.watch() ...
  }
}
```

which generates:

```
abstract class _$ActiveFoos
    extends BuildlessAutoDisposeStreamNotifier<Set<Foo>> {
  late final AutoDisposeRef<dynamic> ref;  // <-- the issue is with 'ref'

  Stream<Set<Foo>> build(
    AutoDisposeRef<dynamic> ref,
  );
}
```

I would like to point out that I have successfully built a number of provider chains (one provider watching another and being watched by another) before attempting to switch to codegen, so this feels a little bit like a small nitpick. Riverpod is a fantastic package and is central to my app's state management. Thank you!

**Comment by rrousselGit** [comment, marked as answer]
Your `build` usage is incorrect. Don't pass "ref" as parameter there. Do `Stream<T> build()`

**Comment by 2x2xplz** [comment]
OK I got it now, just plain `build()` without parameters when using the class syntax, `build(MyFunctionRef ref)` when using the functional syntax. The name of the parameter is the same as the function name + Ref. Working well now, thank you.
